lib/sqlite3/statement.rb

```diff
@@ -145,6 +145,29 @@ def must_be_open! # :nodoc:
       end
     end
 
+    # Returns a Hash containing information about the statement.
+    # The contents of the hash are implementation specific and may change in
+    # the future without notice. The hash includes information about internal
+    # statistics about the statement such as:
+    #   - +fullscan_steps+: the number of times that SQLite has stepped forward
+    # in a table as part of a full table scan
+    #   - +sorts+: the number of sort operations that have occurred
+    #   - +autoindexes+: the number of rows inserted into transient indices
+    # that were created automatically in order to help joins run faster
+    #   - +vm_steps+: the number of virtual machine operations executed by the
+    # prepared statement
+    #   - +reprepares+: the number of times that the prepare statement has been
+    # automatically regenerated due to schema changes or changes to bound
+    # parameters that might affect the query plan
+    #   - +runs+: the number of times that the prepared statement has been run
+    #   - +filter_misses+: the number of times that the Bloom filter returned
+    # a find, and thus the join step had to be processed as normal
+    #   - +filter_hits+: the number of times that a join step was bypassed
+    # because a Bloom filter returned not-found
+    def stat hash_or_key = nil
+      stmt_stat hash_or_key
```

Comment:
Just a suggestion: if we implement 2 private methods in C then maybe we could pull branching logic in to Ruby.

For example:

```ruby
def stat key = nil
  if key
    stat_for(key)
  else
    stats_as_hash
  end
end
```

Where `stats_as_hash` and `stat_for_key` are private methods implemented in C.  WDYT?

Reply:
I like that. I started off by just trying to understand how the `GC.stat` method was implemented and oriented myself around that. However, once I realized I needed an optional argument and couldn't figure out how to do that in C and so make the Ruby method, I agree that simplifying the C code to just have two separate methods makes better sense. Will do that now

Reply:
Done in 089d00e